src/arith/iter_affine_map.cc

```diff
@@ -1260,140 +1300,132 @@ IterSumExpr IterMapRewriter::PreprocessDividend(IterMapExpr dividend) {
-  info.left_pad = left_pad;
-  info.right_pad = right_pad;
-
-  auto left_padding_introduced = (left_pad != 0);
-  // Equivalent to (0 <= split < left_pad), but easier to simplify in
-  // terms of the transformed variables.
-  auto left_padding_predicate =
-      left_padding_introduced && (floordiv(info.padded, divisor) == floordiv(base, divisor) &&
-                                  floormod(info.padded, divisor) < left_pad);
-
-  PrimExpr nparts = ceildiv(right_edge, divisor);
-
-  auto right_padding_introduced = (right_pad != 0);
-
-  // Equivalent to (right_edge <= split < right_edge+right_pad), but
-  // easier to simplify in terms of the transformed variables.
-  auto right_padding_predicate = right_padding_introduced &&
-                                 (floordiv(info.padded, divisor) == floordiv(right_edge, divisor) &&
-                                  floormod(info.padded, divisor) >= floormod(right_edge, divisor));
-
-  requires_padding_ = requires_padding_ || (left_padding_introduced || right_padding_introduced);
-  padding_predicate_ = padding_predicate_ || (left_padding_predicate || right_padding_predicate);
-
-  return {info.padded, left_pad};
+  if (!info.padded.defined()) {
+    PrimExpr mark_left_pad = info.left_pad.defined() ? info.left_pad : 0;
+    PrimExpr mark_right_pad;
+    if (CanProveDivisible(mark->extent + mark_left_pad, info.padding_factor)) {
+      mark_right_pad = 0;
+    } else {
+      mark_right_pad = floormod(-(mark->extent + mark_left_pad), info.padding_factor);
+    }
+    PrimExpr padded_extent = analyzer_->Simplify(mark_left_pad + mark->extent + mark_right_pad);
+    info.right_pad = mark_right_pad;
+    info.padded = IterMark(IterSumExpr({IterSplitExpr(mark)}, mark_left_pad), padded_extent);
+
+    auto left_padding_introduced = (mark_left_pad != 0);
+    PrimExpr divisor = info.padding_factor;
+    PrimExpr right_edge = mark_left_pad + mark->extent;
+
+    // Equivalent to (0 <= split < left_pad), but easier to simplify in
+    // terms of the transformed variables.
+    auto left_padding_predicate =
+        left_padding_introduced && (floordiv(info.padded->source, divisor) == 0 &&
+                                    floormod(info.padded->source, divisor) < mark_left_pad);
+
+    auto right_padding_introduced = (mark_right_pad != 0);
+
+    // Equivalent to (right_edge <= split < right_edge+right_pad), but
+    // easier to simplify in terms of the transformed variables.
+    auto right_padding_predicate =
+        right_padding_introduced &&
+        (floordiv(info.padded->source, divisor) == floordiv(right_edge, divisor) &&
+         floormod(info.padded->source, divisor) >= floormod(right_edge, divisor));
+
+    requires_padding_ = requires_padding_ || (left_padding_introduced || right_padding_introduced);
+    padding_predicate_ = padding_predicate_ || (left_padding_predicate || right_padding_predicate);
+  }
+  // ICHECK(CanProveDivisible(info.padded->extent, split->lower_factor));
```

Comment:
Would like to check the padding factor is divisible by `split->lower_factor`, then the commented check can be ensured from context. I found it may fail unfortunetely due to simplifier's ability limitation when the padded extent contain complex flm/fld expressions.

Reply:
Got it.  I noticed that there were also some simplification steps that needed to increase the number of iterations performed.  Is the failure to prove divisibility related, since `CanProveDivisible` only uses the default of 2 steps?

(I'm also wondering if the default for `Analyzer::Simplify` should be to iterate until it the simplification converges, rather than using a fixed number of steps.)